### ISSUE TITLE
Check in .tool-versions?

### ISSUE THREAD

**Issue by agjohnson** [readthedocs/readthedocs.org#9516]
This is maybe a broader conversation for all of our repos, but does it make sense to check in `.tool-versions` to this/other repositories? Seems we're moving towards `asdf`, and I'm a fan of the tool. This would provide a way for use to specify the exact version of Python/Node/etc that everyone should use, perhaps even matching up to what we install inside Docker. Folks not using `asdf` won't have any change.

I think one side effect might be pulling `main` in and all of a sudden you need to reinstall Python, that might be annoying. Or I'm sure collecting a pile of unused Python/Node versions as we switch around.

ping @readthedocs/backend and I know @humitos uses `asdf` locally as well

**Comment by humitos** [comment]
In general, I think `.tool-versions` is more useful on projects that do not use Docker.

However, even in `readthedocs.org` project we need Python to install the `requirements/docker.txt` dependencies to be able to launch the Docker setup. In this case, `.tool-versions` could help to auto-enable the correct Python version (in this case, `3.10.x`) --which users _must_ have installed the _exact_ patch version to work.

Another potential problem I see with this auto-setup is that without a `.envrc` file defining:

```
use asdf python
layout python
```

to create a virtual environment for this particular directory/project,  users will install the `requirements/docker.txt` dependencies globally and probably break other things.

**Comment by ericholscher** [comment]
I don't really know what `.tool-versions` does. Are there docs on what this would achieve? Is it `asdf`'s lockfile equivalent? 

In general I'm a bigger fan of Docker, since it seems more widely used in the ecosystem for specifying this stuff. But if we're going to define asdf versions, we should document how to get setup with them nicely.

**Comment by agjohnson** [comment]
> Another potential problem I see with this auto-setup is that without a .envrc

Ahhh yeah, that's a really good point. I do rely on `layout python`, using `.envrc`. I wish there was a global way to set this, as I almost always want this or `layout poetry`.

> Is it asdf's lockfile equivalent?

Fairly close, yeah. We aren't checking these in currently, everyone using `asdf` locally has their own copy, which is in `.gitignore`.

`asdf` replaces `pyenv`, `nodenv`, `rbenv`, etc. It brings together all of the various `.python-version`/`.node-version` files into a single `.tool-versions` file.

> Are there docs on what this would achieve?

Here's an example of setting a local version in a repo:

https://asdf-vm.com/guide/getting-started.html#local

> In general I'm a bigger fan of Docker

Docker is a very heavy solution though. If we're saying that all of our dependencies must be installed in Docker, and there are no dependencies install on our host systems, that is going to break a lot of editor tooling/integrations that rely on host installed dependencies, Python, and Node.

So, we still need some solution to define what is installed and used on the host system anyways. Introducing Docker to other repositories doesn't solve this issue, and so we mostly just end up with two disparate implementations of dependencies.

**Comment by ssbarnea** [comment]
More recently this hardcoded use of asdf with `.tool-versions` become a liability because asdf itself does not support `latest` label but other tools like mise or github setup actions support it:

```
# .tool-versions
nodejs latest
python latest
```

Basically this file chokes on RTD while working for other users and AFAIK there is no way to tell RTD not to use it.

It would be much better if RTD would swap asdf with mise, as it has much better support while still being compatible.

**Comment by agjohnson** [comment]
@ssbarnea Is this an issue you're hitting within a project build? I haven't heard of any similar issue, this would be good to raise as a separate issue if so.

This issue is more around the development use of asdf in our local environments. We use this to manage python/nodejs versions for actually working with the RTD codebase.

**Comment by ericholscher** [comment]
I looked into the tool-versions stuff, and seems useful for us to standardize on asdf and a specific set of versions. Though with docker most of this is defined externally, it's useful for running local tooling, but I do wonder how much we should be running locally these days vs. always inside docker given the new security context.

**Comment by agjohnson** [comment]
It's not a bad way to add additional layers of security, we could add this as a supplementary layer to some of our other repositories. But _requiring_ docker containers for basic development is a big headache. Containers are good are encapsulating running processes but are garbage at nicely abstracting the development process.

Wrapping a container around all development breaks a lot and everything is harder. Not having dependencies is a problem for IDEs being able to lint/test/format/refactor, and everything from copying files, debugging why docker build failed, and having shell history is harder in a container.

If we did add another, _optional_ layer of docker containerization, I'd _still_ use asdf to manage dependencies inside/outside though, so wrapping in docker feels like a separate conversation.

Running inside a docker container would require reinstalling dependencies more aggressively, so at least there wouldn't be the frustration of having your env break. But there would be the frustration of having your env constantly rebuild still.

The most important question I posed in 

- https://github.com/readthedocs/meta/issues/226

is **will having the environment suddenly go bad be too annoying?**

That is:

- I pull the repo
- I install Python 3.12 with `asdf install`
- I install packages with uv
- Someone merges a change to the tool-version file upgrading Python
- I pull `main`
- Immediately my env is broken. I don't have the correct Python installed, none of my dependencies are installed for that Python
- If I remember the previous Python, I can run `asdf set python 3.12.x`

**Comment by humitos** [comment]
> is **will having the environment suddenly go bad be too annoying?**

We will have the same situation for all the NPM work we've done already. We are enforcing a particular version of nodejs there.

> If I remember the previous Python, I can run `asdf set python 3.12.x`

Even if you don't remember it, there is always `git log` at hand 😉 

----

That said, I'm fine when that happens eventually 👍🏼

**Comment by ericholscher** [comment]
I don't think we upgrade Python that frequently, so I'm not too worried about that specifically. But if we started bumping for minor versions, that might get annoying.

**Comment by agjohnson** [comment]
> We will have the same situation for all the NPM work we've done already. We are enforcing a particular version of nodejs there.

Sort of. With NPM we're enforcing a _minimum version_. If you now upgrade locally to 24.17 it's not going to break your env enforcing 24.16 minimum. I wouldn't intend to bump this version up until it's necessary though.

With asdf, selector ranges aren't supported and it's always an explicit version. If we bump that to 24.16.1, you will need to reinstall Node and sometimes may need to do an `npm install` pass after.

> Even if you don't remember it, there is always git log at hand 😉

Yeah there are options, it's just bound to be frustrating. But having said that, we don't upgrade versions super frequently either.

---

I'm also okay with either direction too. We could just try and see if we want to experiment.

Ulteimately, I lean towards being flexible and having a defined, consistent pattern though:

- Node/npm enforces an absolute minimum version
- Python or uv sync enforces an absolute minimum version
- Repositories have some source for suggested versions -- tool version file, env file with asdf versions env vars (`ASDF_PYTHON_VERSION`), direnv include
- Something warns when suggested tool versions are newer -- pre-commit?
- Bonus points: all the places we define versions use a common source of truth -- CircleCI, tox, Dockerfiles

There's probably a minimal pattern here though.
